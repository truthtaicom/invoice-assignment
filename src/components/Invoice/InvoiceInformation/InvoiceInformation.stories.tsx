import { storiesOf } from '@storybook/react';
import React from 'react';
import { action } from '@storybook/addon-actions';
import InvoiceInformation from "./InvoiceInformation"

storiesOf("Component", module)
  .add("Invoice Information", () => (
    <InvoiceInformation date={new Date()} />
  ))