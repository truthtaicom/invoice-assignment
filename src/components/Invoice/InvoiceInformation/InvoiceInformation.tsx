import React, { useState, useEffect } from 'react'
import styled from 'styled-components';
import { Input, DatePicker, ToggleSwitch, Form, FormItem, Button } from '../../../ui-kit'


const StyledInvoiceInformation = styled.div`
  display: grid;
  width: auto;
  padding: 1rem;
`

const StyledButtonSubmit = styled(Button)`
  display: grid;
  margin: 0 auto;
  width: 100%;
  margin-top: 1rem;
`

const InvoiceInformation = React.memo(({ item = { amount: 0, subject: '' }, ...props }: any) => {
  const { date: itemDate } = item
  const [date, setDate] = useState(itemDate ? new Date(itemDate) : new Date())
  const [isRetrieve, setIsRetrive] = useState(false)
  const [amount, setAmount] = useState(item.amount)
  const [subject, setSubject] = useState(item.subject)

  useEffect(() => {
    setAmount(item.amount)
    setSubject(item.subject)
  }, [item.amount, item.subject])

  const datePickerOnchange = (value) => { setDate(value) }
  
  const toggleSwitchOnchange = (value) => {
    setIsRetrive(value)
    props.onRetriveFromBankAcc(value)
  }
  
  
  const subjectOnchange = (e) => { setSubject(e.target.value) }
  const amountOnChange = (e) => { setAmount(e.target.value) }
  const onSubmit = (e) => {
    const values = Object.keys(e).filter(e => e).reduce((prev, cur) => ({ ...prev, [cur]: e[cur] }), { amount: 0 }) || {}
    const amountFromValue = values && values.amount.toString()
    if(isRetrieve) {
      props.onNext({ ...values })
    } else {
      props.onSubmit({ ...values, amount: parseInt(amountFromValue, 10)})
    }
  }

  const isValidWithII = subject && subject.length  && amount
  const isValidWithPayments = subject && subject.length && isRetrieve

  return (
    <StyledInvoiceInformation>
      <Form onSubmit={onSubmit}>
        <FormItem>
          <DatePicker
            name="date"
            label="Date"
            onChange={datePickerOnchange}
            selectedDate={date}
            dateFormat="yyyy-MM-dd"
          />
        </FormItem>
        <FormItem>
          <Input label="Subject" name="subject" defaultValue={subject} onChange={subjectOnchange}/>
        </FormItem>
        <FormItem>
          <ToggleSwitch 
            name="retriveFromBankAcc"
            // isOpen={retriveFromBankAcc}
            label="Retrieve amount from your bank account"
            onChange={toggleSwitchOnchange}
          />
        </FormItem>
        <FormItem>
          { !isRetrieve && <Input name="amount" label="Amount" type="number" unit="EUR" isFormat value={amount} onChange={amountOnChange} /> }
        </FormItem>
        { 
          isRetrieve 
          ? <StyledButtonSubmit color="primary" disabled={!isValidWithPayments}>Next</StyledButtonSubmit>
          : <StyledButtonSubmit color="primary" type="submit" disabled={!isValidWithII}>Done</StyledButtonSubmit>
        }
      </Form>
    </StyledInvoiceInformation>
  )
})

export default InvoiceInformation
