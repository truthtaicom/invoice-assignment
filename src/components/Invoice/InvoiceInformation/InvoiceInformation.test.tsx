import React from 'react'
import { render } from 'react-testing-library'
import InvoiceInformation from './InvoiceInformation'

describe('<InvoiceInformation /> spec', () => {
  it('renders the component', () => {
    const component = render(<InvoiceInformation onRetriveFromBankAcc={jest.fn} />)
    expect(component).toMatchSnapshot()
  })
});