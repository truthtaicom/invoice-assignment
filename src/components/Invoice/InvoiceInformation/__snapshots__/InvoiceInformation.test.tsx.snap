// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<InvoiceInformation /> spec renders the component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-kkGfuU ixzkKG"
      >
        <form
          class="sc-eNQAEJ fYukrM"
        >
          <div
            class="sc-hMqMXs dSDiJQ"
          >
            <div
              class="sc-bwzfXH hEUiIL"
            >
              <div
                class="sc-htpNat cWnILl"
              >
                <div
                  class="sc-bxivhb iiUfTC"
                >
                  <span
                    class="sc-gzVnrw gtppSg"
                  >
                    Date
                  </span>
                  <span
                    class="sc-gzVnrw gtppSg"
                  >
                    2019-05-07
                     
                  </span>
                  <span
                    class="sc-gzVnrw gtppSg"
                  >
                    +
                  </span>
                </div>
                <div
                  class="sc-bxivhb evsUlz"
                >
                  <label
                    class="sc-ifAKCX hSWqAs"
                  >
                     
                    <span
                      class="sc-EHOje cZmnwG"
                    >
                      Enter 
                    </span>
                      
                    Date
                  </label>
                  <div
                    class="react-datepicker-wrapper"
                  >
                    <div
                      class="react-datepicker__input-container"
                    >
                      <input
                        class=""
                        name="date"
                        type="text"
                        value="2019-05-08"
                      />
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="sc-hMqMXs dSDiJQ"
          >
            <div
              class="sc-htpNat cWnILl"
            >
              <div
                class="sc-bxivhb iiUfTC"
              >
                <span
                  class="sc-gzVnrw gtppSg"
                >
                  Subject
                </span>
                <span
                  class="sc-gzVnrw gtppSg"
                >
                  
                   
                </span>
                <span
                  class="sc-gzVnrw gtppSg"
                >
                  +
                </span>
              </div>
              <div
                class="sc-bxivhb evsUlz"
              >
                <label
                  class="sc-ifAKCX hSWqAs"
                >
                   
                  <span
                    class="sc-EHOje cZmnwG"
                  >
                    Enter 
                  </span>
                    
                  Subject
                </label>
                <input
                  class="sc-bZQynM htDgzj"
                  name="subject"
                  value=""
                />
              </div>
            </div>
          </div>
          <div
            class="sc-hMqMXs dSDiJQ"
          >
            <div
              class="sc-jzJRlG iyDAyt"
            >
              <label
                class="sc-cSHVUG rrLZq"
              >
                Retrieve amount from your bank account
              </label>
              <label
                class="sc-kAzzGY gpDVKR"
              >
                <input
                  class="sc-chPdSV gEoWmc"
                  type="checkbox"
                />
                <span
                  class="sc-kgoBCf bAsMgC"
                />
              </label>
            </div>
          </div>
          <div
            class="sc-hMqMXs dSDiJQ"
          >
            <div
              class="sc-htpNat cWnILl"
            >
              <div
                class="sc-bxivhb iiUfTC"
              >
                <span
                  class="sc-gzVnrw gtppSg"
                >
                  Amount
                </span>
                <span
                  class="sc-gzVnrw gtppSg"
                >
                   
                  EUR
                </span>
                <span
                  class="sc-gzVnrw gtppSg"
                >
                  +
                </span>
              </div>
              <div
                class="sc-bxivhb evsUlz"
              >
                <label
                  class="sc-ifAKCX hSWqAs"
                >
                   
                  <span
                    class="sc-EHOje cZmnwG"
                  >
                    Enter 
                  </span>
                    
                  Amount
                </label>
                <input
                  class="sc-bZQynM htDgzj"
                  name="amount"
                  type="number"
                  value="0"
                />
              </div>
            </div>
          </div>
          <button
            class="sc-bdVaJa sc-iAyFgw ixJdnN"
            color="primary"
            disabled=""
            type="submit"
          >
            Done
          </button>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-kkGfuU ixzkKG"
    >
      <form
        class="sc-eNQAEJ fYukrM"
      >
        <div
          class="sc-hMqMXs dSDiJQ"
        >
          <div
            class="sc-bwzfXH hEUiIL"
          >
            <div
              class="sc-htpNat cWnILl"
            >
              <div
                class="sc-bxivhb iiUfTC"
              >
                <span
                  class="sc-gzVnrw gtppSg"
                >
                  Date
                </span>
                <span
                  class="sc-gzVnrw gtppSg"
                >
                  2019-05-07
                   
                </span>
                <span
                  class="sc-gzVnrw gtppSg"
                >
                  +
                </span>
              </div>
              <div
                class="sc-bxivhb evsUlz"
              >
                <label
                  class="sc-ifAKCX hSWqAs"
                >
                   
                  <span
                    class="sc-EHOje cZmnwG"
                  >
                    Enter 
                  </span>
                    
                  Date
                </label>
                <div
                  class="react-datepicker-wrapper"
                >
                  <div
                    class="react-datepicker__input-container"
                  >
                    <input
                      class=""
                      name="date"
                      type="text"
                      value="2019-05-08"
                    />
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="sc-hMqMXs dSDiJQ"
        >
          <div
            class="sc-htpNat cWnILl"
          >
            <div
              class="sc-bxivhb iiUfTC"
            >
              <span
                class="sc-gzVnrw gtppSg"
              >
                Subject
              </span>
              <span
                class="sc-gzVnrw gtppSg"
              >
                
                 
              </span>
              <span
                class="sc-gzVnrw gtppSg"
              >
                +
              </span>
            </div>
            <div
              class="sc-bxivhb evsUlz"
            >
              <label
                class="sc-ifAKCX hSWqAs"
              >
                 
                <span
                  class="sc-EHOje cZmnwG"
                >
                  Enter 
                </span>
                  
                Subject
              </label>
              <input
                class="sc-bZQynM htDgzj"
                name="subject"
                value=""
              />
            </div>
          </div>
        </div>
        <div
          class="sc-hMqMXs dSDiJQ"
        >
          <div
            class="sc-jzJRlG iyDAyt"
          >
            <label
              class="sc-cSHVUG rrLZq"
            >
              Retrieve amount from your bank account
            </label>
            <label
              class="sc-kAzzGY gpDVKR"
            >
              <input
                class="sc-chPdSV gEoWmc"
                type="checkbox"
              />
              <span
                class="sc-kgoBCf bAsMgC"
              />
            </label>
          </div>
        </div>
        <div
          class="sc-hMqMXs dSDiJQ"
        >
          <div
            class="sc-htpNat cWnILl"
          >
            <div
              class="sc-bxivhb iiUfTC"
            >
              <span
                class="sc-gzVnrw gtppSg"
              >
                Amount
              </span>
              <span
                class="sc-gzVnrw gtppSg"
              >
                 
                EUR
              </span>
              <span
                class="sc-gzVnrw gtppSg"
              >
                +
              </span>
            </div>
            <div
              class="sc-bxivhb evsUlz"
            >
              <label
                class="sc-ifAKCX hSWqAs"
              >
                 
                <span
                  class="sc-EHOje cZmnwG"
                >
                  Enter 
                </span>
                  
                Amount
              </label>
              <input
                class="sc-bZQynM htDgzj"
                name="amount"
                type="number"
                value="0"
              />
            </div>
          </div>
        </div>
        <button
          class="sc-bdVaJa sc-iAyFgw ixJdnN"
          color="primary"
          disabled=""
          type="submit"
        >
          Done
        </button>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
